spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/jdbc/DataSourcePoolMetricsAutoConfiguration.java

```diff
@@ -69,10 +74,31 @@
 
 		@Autowired
 		void bindDataSourcesToRegistry(Map<String, DataSource> dataSources, MeterRegistry registry,
-				ObjectProvider<DataSourcePoolMetadataProvider> metadataProviders) {
+				ObjectProvider<DataSourcePoolMetadataProvider> metadataProviders,
+				DataSourcePoolMetricsProperties properties) {
 			List<DataSourcePoolMetadataProvider> metadataProvidersList = metadataProviders.stream()
 					.collect(Collectors.toList());
-			dataSources.forEach(
+
+			// Possibly expand routing datasources into their resolved datasources
+			Map<String, DataSource> allDataSources = new HashMap<>();
+			dataSources.forEach((beanName, dataSource) -> {
+				if (!(dataSource instanceof AbstractRoutingDataSource)) {
+					allDataSources.put(beanName, dataSource);
+					return;
+				}
+				if (properties.isIgnoreRoutingDataSources()) {
+					return;
+				}
+				AbstractRoutingDataSource routingDataSource = (AbstractRoutingDataSource) dataSource;
+				routingDataSource.getResolvedDataSources().forEach((routingKey, routedDataSource) -> {
+					boolean alreadyRegistered = dataSources.containsValue(routedDataSource);
+					if (!properties.isDeduplicateRoutingDataSources() || !alreadyRegistered) {
+						allDataSources.put(getRoutedDataSourceName(beanName, routingKey), routedDataSource);
```

Comment:
This is tricky.

If a `DataSource` is a target of multiple routing data sources, the name for its metrics will depend upon the ordering of the `dataSources` map. Similarly, if a `DataSource` is both a top-level bean and a target of an `AbstractRoutingDataSource` its name will also depend upon the ordering of the `dataSources` map.

I'm not sure exactly what the answer to this is at the moment, but I am sure that the names need to be deterministic. While the injected `dataSources` map is a `LinkedHashMap` (so it'll iterate in insertion order), I'm not sure that Framework guarantees that order. My reading of `DefaultListableBeanFactory.findAutowireCandidates` is that it does not.

Let's take a step back until we've completed the design work on #22823.